### PR TITLE
Start building pre-releases with cythonized scheduler

### PR DESCRIPTION
**Pull request by charlesbluca**
This PR modifies the Distributed pre-release recipe to build Cython variants of Distributed using the steps outlined by @jakirkham in https://github.com/dask/distributed/issues/4442#issuecomment-763074600.

Choosing which Distributed variant to install should be doable by specifying the `distributed-proc` build string during an install / env creation command, though with `track_features="cythonized-scheduler"` specified for Cython builds Conda should always prioritize non-Cython builds by default.

- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by jrbourbeau — COMMENTED** [review]
Thanks @charlesbluca. There was some talk during the last monthly community meeting of dropping the cythonized version of the scheduler (also xref https://github.com/dask/distributed/issues/5685). My guess is we should wait for a decision on that before adding extra builds here

**Comment by jakirkham** [comment]
James as part of that discussion we had decided that if some work was done here to setup nightlies and refactor out the Cythonized parts, we could keep it. This is one of those steps. However I think it is important for us to have the opportunity to progress here otherwise none of the actionables requested can complete. IOW it becomes a Catch 22

**Comment by jakirkham** [comment]
Right this was one of the issues Florian had raised in the issue above and was discussed in the meeting. So this is our attempt to address that.

Agree there is still refactoring work to do as well.

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:12]: Normally in conda-forge we have used `-split` to indicate split output

```suggestion
  name: distributed-split
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:52]: Would suggest we skip this for now. We could add if we wanted to cross-compile for ARM or PPC, but don't think there is a need for that atm

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:54]: Curious about the version constraint here. How was this decided?

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:73]: As we just dropped Python 3.7, think we can simplify this to

```suggestion
        - tornado >=6.0.3
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:43]: Would add `--no-deps` as it looks like `pip` is trying to install things, which it shouldn't do.

Enabling profiling on our nightly packages makes sense. Just noting that we could also disable profiling if needed

```suggestion
      script: python -m pip install . -vv --no-deps --install-option="--with-cython=profile"  # [cython_enabled]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:44]: Would suggest dropping the square brackets. Otherwise this seems good

```suggestion
      track_features: {{ "cythonized-scheduler" if cython_enabled else "" }}
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:90]: Sorry for the stray comment here

Maybe it is worth having some kind of test to ensure we got the Cythonized version or not?

```suggestion
        - dask-worker --help
        - python -c "from distributed.scheduler import COMPILED; assert COMPILED is {{ cython_enabled }}"
```

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/conda_build_config.yaml]: Curious as to what this would do - since we are already building `linux-64` packages, would this just stop the cython builds from being converted to non-Linux platforms?

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/conda_build_config.yaml]: Maybe we should defer this to some other maintainers? cc @jrbourbeau @jsignell

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml:22]: Yeah another name makes more sense here - `impl` sounds good to me but happy to defer this to others as well. Will replace all instances of `proc` for `impl` in a separate commit

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml:25]: I don't think that would cause any issues in particular, though I wonder if it's necessary - unless I'm missing something, we should only really need one published selector package to handle the choice between Cython / no-Cython builds for all pre-release packages, right? Doing this makes it so that we would only upload a new selector package if we decided to do so by bumping the version / build (i.e. if we make some change to how it operates and want to update).

That being said, that does still require us to manually bump the `proc_version/build_number` manually, and the packages themselves are also very small (all `arrow-cpp-proc` packages are under 30kB each) so I guess outside of some redundancy it wouldn't be an issue to host them.

Happy to make the switch if you have other opinions on this though 🙂

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml:54]: I just picked this arbitrarily because RAPIDS projects used it - meant to put a review comment asking what this should actually be 😅

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml]: Anything else we should do to follow up on this so that we're only building for a single Python version?

**Review by charlesbluca — COMMENTED** [review]
Thanks for the thorough review @jakirkham 😄 a few responses:

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [reply continuous_integration/recipes/distributed/meta.yaml:25]: Mostly was seeing this as a simplification. Fewer things to maintain.

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (jakirkham) [reply continuous_integration/recipes/distributed/conda_build_config.yaml]: Sure happy to hear other thoughts here 🙂

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [reply continuous_integration/recipes/distributed/meta.yaml]: We might want to add `noarch: python  # [nocython]` so that we get `noarch` packages for pure Python. Then we can drop the conversion step as well.

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:26]: As this should be architecture agnostic, we can use `noarch` here

```suggestion
      string: {{ build_ext }}
      noarch: generic
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:41]: Suggest enabling `noarch` for pure Python builds

```suggestion
      string: py{{ python | replace(".", "") }}_{{ GIT_DESCRIBE_HASH }}_{{ GIT_DESCRIBE_NUMBER }}_{{ build_ext }}
      noarch: python # [nocython]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:26]: As this should be architecture agnostic, we can use `noarch` here

```suggestion
      string: {{ build_ext }}
      noarch: generic
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [reply continuous_integration/recipes/distributed/meta.yaml:52]:
```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Added some suggestions above to make this `noarch`. We should be able to drop [the `conda convert` lines]( https://github.com/dask/distributed/blob/d2d76c0ea9737586fe122b5244a873a42cd4787a/.github/workflows/conda.yml#L70-L76 ) with this

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:57]:
```suggestion
        - pip
        - cython                                 # [cython_enabled]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [reply continuous_integration/recipes/distributed/meta.yaml:54]: Sorry should have added this goes in `host`

```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Sorry for the 2nd review batch. Thinking this will resolve the CI issue

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml:25]: Sure, I'm fine with consolidating the version / number here then!

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/conda_build_config.yaml]: Cool, just interested in the impact - I'll go ahead and leave this selector here on the off-chance that we ever begin running builds on a non-Linux platform

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/conda_build_config.yaml]: Cool, renaming to that for now

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:42]: Maybe we can consolidate these like so?

```suggestion
      script: |
        python -m pip install . -vv
          --no-deps
          --install-option="--with-cython=profile"  # [build_ext == "cython"]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:5]: Jinja unfortunately doesn't affect selectors. So we would need to perform this comparison in selectors

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:50]:
```suggestion
        - {{ compiler('c') }}                    # [build_ext == "cython"]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:54]:
```suggestion
        - cython                                 # [build_ext == "cython"]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:40]:
```suggestion
      noarch: python                                                                          # [build_ext == "cython"]
```

**Review by jakirkham — COMMENTED** [review]
This is looking pretty good. Thanks for all of the work here Charles 😄

Had a couple suggestions related to how selectors and Jinja work. IIUC this should also fix the CI issue

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml:40]: Think this one should be

```suggestion
      noarch: python                                                                          # [build_ext == "python"]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml:5]: Ah thanks for the clarification here - good to know for the future 🙂

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:41]: Oops 🤦‍♂️ Sorry forgot we want this to combine onto a single line

```suggestion
      script: >
```

Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml]: No worries - thanks for the quick follow up!

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:44]: Was just about to suggest this 😄

Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml]: Followed up with a small consolidation of this in https://github.com/dask/distributed/pull/5831/commits/7866647fc7f28267142c2d09a3626ceb896e5e30, but it looks like this got builds passing 🎉

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/conda_build_config.yaml:6]:
```suggestion
cython_enabled:
    - True   # [linux]
    - False
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:5]:
```suggestion
{% set build_ext = "cython" if cython_enabled else "python" %}
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:40]:
```suggestion
      noarch: python                              # [not cython_enabled]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:43]:
```suggestion
        --install-option="--with-cython=profile"  # [cython_enabled]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:51]:
```suggestion
        - {{ compiler('c') }}                     # [cython_enabled]
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:55]:
```suggestion
        - cython                                  # [cython_enabled]
```

**Review by jakirkham — COMMENTED** [review]
Not that we need to do this, but this is another way we could handle the Jinja/selector Cython portion

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (charlesbluca) [inline continuous_integration/recipes/distributed/conda_build_config.yaml:3]: Should we remove this now that Python version is negligible?

```suggestion
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [reply continuous_integration/recipes/distributed/meta.yaml:5]: Ofc. Showed another way we could do this in this review ( https://github.com/dask/distributed/pull/5831#pullrequestreview-890316547 ). Though ok with either approach 🙂

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [inline continuous_integration/recipes/distributed/meta.yaml:39]: If so we could also remove python version from the build string:

```suggestion
      string: py_{{ GIT_DESCRIBE_HASH }}_{{ GIT_DESCRIBE_NUMBER }}_{{ build_ext }}
```

**Review by charlesbluca — COMMENTED** [review]
Thanks for all the help here @jakirkham 😄 some questions w.r.t. shifting towards `noarch`, as I notice we are still building py38 / py39 packages:

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/conda_build_config.yaml:6]: Yeah I like this handling, will commit these suggestions

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:40]: Would skip any Python versions before those we support

```suggestion
      noarch: python                              # [not cython_enabled]
      skip: True                                  # [cython_enabled and py<38]
```

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (jakirkham) [reply continuous_integration/recipes/distributed/conda_build_config.yaml:3]: Possibly if we add a `skip`

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [reply continuous_integration/recipes/distributed/meta.yaml:39]: We would need to adjust this to handle the Cython case where this is still needed. Perhaps with a selector?

**Review by jakirkham — COMMENTED** [review]
The different versions are still needed for Cython (though not pure Python). That said, maybe there is a way to simplify things if we use a `skip`. Would also want to check what Python versions are built against by default with Conda-Build

**Comment by charlesbluca** [comment]
Hmm looks like something in https://github.com/dask/distributed/pull/5831/commits/27f826bd56427d74be90dbbad8b5b00670dd5846 caused us to skip over building the `python` variants?

```
 INFO :: The inputs making up the hashes for the built packages are as follows:
{
  "distributed-2022.02.1a220222-py38_gce58db5e_44_cython": {
    "recipe": {
      "cython_enabled": "False"
    }
  },
  "distributed-2022.02.1a220222-py39_gce58db5e_44_cython": {
    "recipe": {
      "cython_enabled": "False"
    }
  },
  "distributed-impl-2022.02.1a220222-cython": {
    "recipe": {}
  }
}
```

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/conda_build_config.yaml:6]: This shouldn't matter, but maybe I'm missing something

https://yaml.org/type/bool.html

File: continuous_integration/recipes/distributed/meta.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/meta.yaml:5]: Maybe the ternary expression isn't working as expected?

```suggestion
{% set build_ext = "cython" %}  # [cython_enabled]
{% set build_ext = "python" %}  # [not cython_enabled]
```

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (charlesbluca) [reply continuous_integration/recipes/distributed/conda_build_config.yaml:6]: Yeah that's what I figured, made the change because the syntax highlighting on GitHub's end made it seem like the two variants were being parsed differently

Comment (jakirkham) [reply continuous_integration/recipes/distributed/conda_build_config.yaml]: Ah ok. Good to know. Though a bit odd they make it to Jinja without being interpreted as `bool`s here first

**Comment by jakirkham** [comment]
Looks like it is doing the right thing now

<details>

```
INFO :: The inputs making up the hashes for the built packages are as follows:
{
  "distributed-2022.02.1a220222-py38_g293ef607_47_cython": {
    "recipe": {
      "c_compiler": "gcc",
      "cython_enabled": "True",
      "target_platform": "linux-64"
    }
  },
  "distributed-2022.02.1a220222-py38_g293ef607_47_python": {
    "recipe": {
      "cython_enabled": "False"
    }
  },
  "distributed-2022.02.1a220222-py39_g293ef607_47_cython": {
    "recipe": {
      "c_compiler": "gcc",
      "cython_enabled": "True",
      "target_platform": "linux-64"
    }
  },
  "distributed-2022.02.1a220222-py39_g293ef607_47_python": {
    "recipe": {
      "cython_enabled": "False"
    }
  },
  "distributed-impl-2022.02.1a220222-cython": {
    "recipe": {}
  },
  "distributed-impl-2022.02.1a220222-python": {
    "recipe": {}
  }
}
```

</details>

File: continuous_integration/recipes/distributed/conda_build_config.yaml
Comment (jakirkham) [inline continuous_integration/recipes/distributed/conda_build_config.yaml:3]: It looks like this is no longer building for Python 3.9. Only Python 3.8. So I think we need to add this back

<details>

```
INFO :: The inputs making up the hashes for the built packages are as follows:
{
  "distributed-2022.02.1a220222-py38_ge7da29f6_50_cython": {
    "recipe": {
      "c_compiler": "gcc",
      "cython_enabled": "True",
      "target_platform": "linux-64"
    }
  },
  "distributed-2022.02.1a220222-py_ge7da29f6_50_python": {
    "recipe": {
      "cython_enabled": "False"
    }
  },
  "distributed-impl-2022.02.1a220222-cython": {
    "recipe": {}
  },
  "distributed-impl-2022.02.1a220222-python": {
    "recipe": {}
  }
}
```

</details>

**Review by jakirkham — APPROVED** [review]
LGTM. Thanks Charles! 😄

**Comment by jakirkham** [comment]
@jrbourbeau @jsignell, do either of you have more thoughts here? 🙂

**Comment by jsignell** [comment]
No thoughts from me. I'm happy if you are happy :)

**Review by jrbourbeau — COMMENTED** [review]
Yep, same here, I'm happy if you're happy with the changes here. 

(This is a non-blocking comment) I noticed the nightly CI build here takes ~10 minutes and is run on all pull requests and pushes to `main`. That seems like a significant amount of CI time to me. I'm wondering how much value there is in building that often vs. only on pushes to `main` or even a single nightly cron job. `dask/dask` and `dask/distributed` have been quite active recently and I've noticed longer wait times for CI builds to start. If we switched to only building on pushes to `main` or a nightly cron job, would folks using these builds miss out on much?

**Comment by jakirkham** [comment]
There is probably still some value in building these on PRs that change the Conda packages in particular (though maybe this can be skipped for other PRs)

Otherwise building on `main` should be fine

**Comment by jrbourbeau** [comment]
> PRs that change the Conda packages in particular (though maybe this can be skipped for other PRs)

Ah, that sounds like a nice balance. It looks like GitHub has support for only triggering builds when certain files are modified https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore. Again, I don't mean for this to block this PR. We can make further updates in future PRs

**Comment by charlesbluca** [comment]
Yeah was about to suggest the same method - I'm okay with pushing a quick commit doing this targetting the following files:

- the conda recipes
- `setup.py`
- the workflow file controlling the pre-release builds

**Comment by jakirkham** [comment]
@jrbourbeau how do the changes above look? 🙂

(thanks Charles for adding those! 🙏)

**Review by jrbourbeau — COMMENTED** [review]
Nice, LGTM -- thanks @charlesbluca

**Comment by jakirkham** [comment]
Thanks Charles for the PR and Julia & James for the reviews! 😄

**Comment by jakirkham** [comment]
Tested the file filtering logic in PR ( https://github.com/dask/distributed/pull/5826 ) and it works as expected (Conda nightly build skipped)! 🎉

**Pull request merged**
